src/analysis/processing/qgsalgorithmrectanglesovalsdiamonds.cpp

```diff
@@ -0,0 +1,209 @@
+  rotationParam->setIsDynamic( true );
+  rotationParam->setDynamicPropertyDefinition( QgsPropertyDefinition( QStringLiteral( "Rotation" ), QObject::tr( "Rotation" ), QgsPropertyDefinition::Double ) );
+  rotationParam->setDynamicLayerParameterName( QStringLiteral( "INPUT" ) );
+  addParameter( rotationParam.release() );
+
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "SEGMENTS" ), QObject::tr( "Segments" ), QgsProcessingParameterNumber::Integer, 5, false, 1 ) );
+}
+
+bool QgsRectanglesOvalsDiamondsAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mShape = parameterAsEnum( parameters, QStringLiteral( "SHAPE" ), context );
+
+  mWidth = parameterAsDouble( parameters, QStringLiteral( "WIDTH" ), context );
+  mDynamicWidth = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "WIDTH" ) );
+  if ( mDynamicWidth )
+    mWidthProperty = parameters.value( QStringLiteral( "WIDTH" ) ).value< QgsProperty >();
+
+  mHeight = parameterAsDouble( parameters, QStringLiteral( "HEIGHT" ), context );
+  mDynamicHeight = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "HEIGHT" ) );
+  if ( mDynamicHeight )
+    mHeightProperty = parameters.value( QStringLiteral( "HEIGHT" ) ).value< QgsProperty >();
+
+  mRotation = parameterAsDouble( parameters, QStringLiteral( "ROTATION" ), context );
+  mDynamicRotation = QgsProcessingParameters::isDynamic( parameters, QStringLiteral( "ROTATION" ) );
+  if ( mDynamicRotation )
+    mRotationProperty = parameters.value( QStringLiteral( "ROTATION" ) ).value< QgsProperty >();
+
+  mSegments = parameterAsDouble( parameters, QStringLiteral( "SEGMENTS" ), context );
+
+  return true;
+}
+
+QgsFeatureList QgsRectanglesOvalsDiamondsAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsFeature outFeature = feature;
+  if ( outFeature.hasGeometry() )
+  {
+    double width = mWidth;
+    if ( mDynamicWidth )
+      width = mWidthProperty.valueAsDouble( context.expressionContext(), width );
+
+    double height = mHeight;
+    if ( mDynamicHeight )
+      height = mHeightProperty.valueAsDouble( context.expressionContext(), height );
+
+    double rotation = mRotation;
+    if ( mDynamicRotation )
+      rotation = mRotationProperty.valueAsDouble( context.expressionContext(), rotation );
+
+    if ( width == 0 || height == 0 )
+    {
+      throw QgsProcessingException( QObject::tr( "Width and height should be greater than 0." ) );
+    }
+
+    double phi = rotation * M_PI / 180;
+    double xOffset = width / 2.0;
+    double yOffset = height / 2.0;
+    QgsPointXY point = outFeature.geometry().asPoint();
+    double x = point.x();
+    double y = point.y();
+
+    QgsPolylineXY ring;
```

Comment:
Instead of using the QgsPolylineXY class (which is slow), you could use the optimised method of creating polygons which is to construct QVector<double> for x and y, and use the QgsLineString( const QVector<double>& x, const QVector<double>& y ) constructor (and then set this as the polygon exterior ring). That's the **fastest** way to create polygons in qgis.

Reply:
Should be fixed